### PR TITLE
An attempt at adding RELAYMSG support to LinkRelay

### LINKED ISSUE #338: LinkRelay: Add support for RELAYMSG
I'm not a big fan of this spec, but some communities are, so let's add configurable support for it as soon as there is a published vendored spec. Currently there is a draft here: https://github.com/ircv3/ircv3-specifications/pull/417

Requirements:

1. published spec (probably vendored)
2. guarantees that receivers don't get a spoofed NUH that could be used to impersonate users
3. ditto with account-tag

### PR DESCRIPTION
**Pull request by necessary129**
Fixes #338

### CODE REVIEW

**Comment by ProgVal** [comment]
Hi,

Without going into the code change itself, I do not want any of the software I maintain to emit RELAYMSG in its current state.

1. The capability name should be vendored, not remain `draft/` forever. https://ircv3.net/specs/extensions/capability-negotiation.html#draft-capabilities
2. The NUH of PRIVMSGs generated from RELAYMSG messages is unspecified, and Ergo's current implementation makes it easy to impersonate the bot
3. Interaction with `account-tag` is also unspecified, and could be used to impersonate the bot

I already made the authors of the RELAYMSG spec aware of these issues, and I will review this PR if and when this is addressed

**Comment by necessary129** [comment]
Aah. ok. I'll close this PR for the moment and will open one when your demands are met.

**Comment by ProgVal** [comment]
On second thought, as the RELAYMSG is unlikely to every get fixed and this is a very requested feature, let's implement it anyway.

File: LinkRelay/config.py [hunk 1]
```diff
@@ -90,6 +90,9 @@ class ValidNonPrivmsgsHandling(registry.OnlySomeStrings):
     registry.Boolean(False, _("""Determines whether the bot will relay NOTICE
     messages.""")))
 
+conf.registerChannelValue(LinkRelay, 'useRelayMsg',
+    registry.Boolean(False, _("""Use RELAYMSG if supported.""")))
+
```
Comment (ProgVal) [inline LinkRelay/config.py:95]: This should mention explicitly the risks involved with using RELAYMSG.

File: LinkRelay/ircmsgs.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    if conf.supybot.protocols.irc.strictRfc():
+        assert (areReceivers(recipient)), repr(recipient)
+        assert s, 's must not be empty.'
+    if minisix.PY2 and isinstance(s, unicode):
+        s = s.encode('utf8')
+    assert isinstance(s, str)
+    if msg and not prefix:
+        prefix = msg.prefix
+    return IrcMsg(prefix=prefix, command='RELAYMSG',
+                  args=(recipient, spoofed_nick, s), msg=msg)
```
Comment (ProgVal) [inline LinkRelay/ircmsgs.py:15]: You can keep this in `plugin.py`

File: LinkRelay/ircmsgs.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+from supybot.ircmsgs import *
+
+
+def relaymsg(recipient, spoofed_nick, s, prefix='', msg=None):
+    """Returns a RELAYMSG to recipient with the message s."""
+    if conf.supybot.protocols.irc.strictRfc():
+        assert (areReceivers(recipient)), repr(recipient)
+        assert s, 's must not be empty.'
+    if minisix.PY2 and isinstance(s, unicode):
+        s = s.encode('utf8')
```
Comment (ProgVal) [inline LinkRelay/ircmsgs.py:10]: Not needed anymore

File: LinkRelay/plugin.py [hunk 4]
```diff
@@ -302,11 +303,19 @@ def send(s):
...
+                    '{}{}{}'.format(args['nick'], seperator, irc.network))
+                    args['draft/relaymsg'] = True
+                    s = args['text']
+                elif isPrivmsg or \
                         self.registryValue('nonPrivmsgs', channel) == 'privmsg':
-                    f = ircmsgs.privmsg
+                    f = functools.partial(ircmsgs.privmsg, relay.targetChannel)
                 elif self.registryValue('nonPrivmsgs', channel) == 'notice':
-                    f = ircmsgs.notice
+                    f = functools.partial(ircmsgs.notice, relay.targetChannel)
```
Comment (ProgVal) [inline LinkRelay/plugin.py:318]: Why this change?

File: LinkRelay/plugin.py [hunk 5]
```diff
@@ -317,9 +326,10 @@ def send(s):
                 tail = [cont + ' ' + s[i:i+remainingLength] for i in
                         range(allowedLength, len(s), remainingLength)]
                 for s in [head] + tail:
-                    msg = f(relay.targetChannel, s)
+                    s = '\x01ACTION {}\x01'.format(s) if args.get('draft/relaymsg') and args.get('action') else s
+                    msg = f(s)
                     msg.tag('relayedMsg')
-                    if 'message-tags' in targetIRC.state.capabilities_ack \
+                    if not args.get('draft/relaymsg') and 'message-tags' in targetIRC.state.capabilities_ack \
```
Comment (ProgVal) [inline LinkRelay/plugin.py:332]: Line wrap

**Review by ProgVal — CHANGES_REQUESTED** [review]
Please gate this feature behind these `supybot.protocols.irc.experimentalExtensions` in addition to the new config var, because this is a draft specification

File: LinkRelay/plugin.py (hunk 4 above)
Comment (necessary129) [reply LinkRelay/plugin.py:318]: This is because the args of `ircmsgs.privmsg` and `ircmsgs.notice` are `recipient` and `s` while that of `relaymsg` are `recipient`, `spoofed_nick` and `s`.

I could reorder the args of the relaymsg function to be `(spoofed_nick, recipient, s)` and then only apply `partial` when the `msg` to be sent is a relaymsg (`f = functools.partial(ircmsgs.relaymsg, sppofed_nick)` and `f = ircmsgs.notice` for others etc) and use the old `msg = f(relay.targetChannel, s)`. But then the order of args would not be same as in the spec (not needed, but it seems cleaner to follow that order)

So it seemed cleaner to me to apply partial all functions to sort out the order and only call `msg = f(s)`.

Comment (ProgVal) [reply LinkRelay/plugin.py:318]: hmm ok then

File: LinkRelay/plugin.py [hunk 6]
```diff
@@ -317,10 +338,12 @@ def send(s):
                 tail = [cont + ' ' + s[i:i+remainingLength] for i in
                         range(allowedLength, len(s), remainingLength)]
                 for s in [head] + tail:
-                    msg = f(relay.targetChannel, s)
+                    s = '\x01ACTION {}\x01'.format(s) if args.get('draft/relaymsg') and args.get('action') else s
```
Comment (ProgVal) [inline LinkRelay/plugin.py:341]:
```
if args.get('draft/relaymsg') and args.get('action'):
    s = '\x01ACTION %s\x01' %s 
```

More readable + doesn't need a line wrap + using percent-substitution for consistency

File: LinkRelay/plugin.py [hunk 7]
```diff
@@ -302,11 +314,20 @@ def send(s):
...
                               (relay.targetChannel, relay.targetNetwork))
             else:
-                if isPrivmsg or \
+                if isPrivmsg and 'draft/relaymsg' in targetIRC.state.capabilities_ls \
+                    and self.registryValue('useRelayMsg', relay.targetChannel) \
+                    and conf.supybot.protocols.irc.experimentalExtensions() \
+                    and targetIRC.state.channels[relay.targetChannel].isOp(targetIRC.nick):
+                    seperator = targetIRC.state.capabilities_ls['draft/relaymsg']
+                    f = functools.partial(relaymsg, relay.targetChannel,
+                    '{}{}{}'.format(args['nick'], seperator, irc.network))
```
Comment (ProgVal) [inline LinkRelay/plugin.py:323]:
```
separator = targetIRC.state.capabilities_ls['draft/relaymsg']
spoofed_nick = ''.join([args['nick'], separator, irc.network)])
f = functools.partial(relaymsg, relay.targetChannel, separator
```

typo fix + uses percent-substitution + gives an explicit name to the computed value + fixes indentation

File: LinkRelay/plugin.py [hunk 8]
```diff
@@ -302,11 +314,20 @@ def send(s):
                 self.log.info('LinkRelay:  I\'m not in in %s on %s' %
                               (relay.targetChannel, relay.targetNetwork))
             else:
-                if isPrivmsg or \
+                if isPrivmsg and 'draft/relaymsg' in targetIRC.state.capabilities_ls \
+                    and self.registryValue('useRelayMsg', relay.targetChannel) \
```
Comment (ProgVal) [inline LinkRelay/plugin.py:318]: Also give the network name to `self.registryValue`

**Review by ProgVal — CHANGES_REQUESTED** [review]

**Review by ProgVal — CHANGES_REQUESTED** [review]
Thanks! One last thing: can you rename `useRelayMsg` to `spoofNicknames`? I think it's clearer for users, who don't care about the detail of the implementation/protocol.

**Comment by ProgVal** [comment]
Thank. I'll sleep on it and merge tomorrow if I don't have another change of heart

**Comment by ProgVal** [comment]
I didn't change my mind!
